h2o-algos/src/main/java/hex/tree/isoforextended/ExtendedIsolationForest.java

```diff
@@ -0,0 +1,377 @@
+            this._seed = _seed;
+            this._extensionLevel = _extensionLevel;
+
+            int maxNumNodesInTree = (int) Math.pow(2, _heightLimit) - 1;
+            this._nodes = new Node[maxNumNodesInTree]; 
+        }
+
+        /**
+         * Implementation of Algorithm 2 (iTree) from paper.
+         * Algorithm not need exact row vectors in scoring (evaluation) stage but just number of rows in node.
+         * Therefore nodeFrames are removed from DKV after filtering.
+         */
+        public void buildTree() {
+            Frame frame = DKV.get(_frameKey).get();
+            _nodes[0] = new Node(frame._key, frame.numRows(), 0);
+            for (int i = 0; i < _nodes.length; i++) {
+                Node node = _nodes[i];
+                if (node == null || node._external) {
+                    continue;
+                }
+                Frame nodeFrame = node.getFrame();
+                int currentHeight = node._height;
+                if (node._height >= _heightLimit || nodeFrame.numRows() <= 1) {
+                    node._external = true;
+                    node._numRows = nodeFrame.numRows();
+                    node._height = currentHeight;
+                    DKV.remove(node._frameKey);
+                } else {
+                    currentHeight++;
+
+                    node._p = VecUtils.uniformDistrFromFrame(nodeFrame, _seed + i);
+                    node._n = ArrayUtils.gaussianVector(
+                            nodeFrame.numCols(), _seed + i, nodeFrame.numCols() - _extensionLevel - 1);
+                    Frame sub = MatrixUtils.subtractionMtv(nodeFrame, node._p);
+                    Vec mul = MatrixUtils.productMtvMath(sub, node._n);
+                    Frame left = new FilterLtTask(mul, 0)
+                            .doAll(nodeFrame.types(), nodeFrame)
+                            .outputFrame(Key.make(), nodeFrame._names, nodeFrame.domains());
+                    Frame right = new FilterGteTask(mul, 0)
+                            .doAll(nodeFrame.types(), nodeFrame)
+                            .outputFrame(Key.make(), nodeFrame._names, nodeFrame.domains());
+                    DKV.remove(nodeFrame._key);
+
+                    if (rightChildIndex(i) < _nodes.length) {
+                        _nodes[leftChildIndex(i)] = new Node(left._key, left.numRows(), currentHeight);
+                        _nodes[rightChildIndex(i)] = new Node(right._key, right.numRows(), currentHeight);
+                    } else {
+                        DKV.remove(left._key);
+                        DKV.remove(right._key);
+                    }
+                }
+            }
+        }
+        
+        private int leftChildIndex(int i) {
+            return 2 * i + 1;
+        }
+
+        private int rightChildIndex(int i) {
+            return 2 * i + 2;
+        }        
+
+        /**
+         * Helper method. Print nodes size of the tree.
+         */
+        public void print() {
```

Comment:
would rename to logNodes() and logLeight()

Reply:
Implemented in PR: https://github.com/h2oai/h2o-3/pull/4907